### ISSUE TITLE
Workflow for fleet managers that don't support specifying path

### ISSUE THREAD

**Issue by rcywongaa** [osrf/rmf_core#124]
As mentioned in https://github.com/osrf/rmf_core/issues/104#issuecomment-643994871, interfacing with a vendor fleet manager involves creating a fleet adapter and fleet driver.

For creating the fleet adapter, one needs to implement [`rmf_fleet_adapter::agv::RobotCommandHandle`](https://github.com/osrf/rmf_core/blob/master/rmf_fleet_adapter/include/rmf_fleet_adapter/agv/RobotCommandHandle.hpp#L70) which includes the function `follow_new_path`.  I assume this function is for `rmf_traffic` to instruct the fleet manager that a specific AGV is to follow a specific path.

How would this work for fleet managers that plan the paths internally and do not support specifying the path?

I suppose this means the fleet adapter, on receiving a call to `follow_new_path`, only sends `DestinationRequest` messages to the fleet driver using only the destination.  Will such discarding of path information affect the internal operation of rmf_core?

**Comment by mxgrey** [comment]
> interfacing with a vendor fleet manager involves creating a fleet adapter and fleet driver.

This used to be the case back in December, but it is no longer the recommended way to interface with RMF (support does still exist for this in the first release, but we will likely drop it by the time RMF is considered ready for production).

Instead the recommended way to interface with a vendor fleet manager is to use the `rmf_fleet_adapter` library (i.e. the `RobotCommandHandle` interface and other related APIs in [this directory](https://github.com/osrf/rmf_core/tree/master/rmf_fleet_adapter/include/rmf_fleet_adapter/agv)) to create your own fleet adapter executable that talks between RMF and the fleet manager. The simple way to think about the difference between this new approach and the old one is that the old approach had two executables:

1. the fleet driver (written by the vendor or system integrator)
2. the fleet adapter (provided out of the box by RMF)

The new approach uses only one executable:

1. the fleet adapter (written by a vendor or system integrator using the APIs provided by RMF)

> How would this work for fleet managers that plan the paths internally and do not support specifying the path?

This depends on what level of path planning the fleet manager can/must do. If we're talking about local planning around environmental obstacles, then I would recommend feeding the fleet manager the waypoints given by `follow_new_path` one at a time as the robot reaches the waypoints. The fleet adapter does not attempt to deal with environmental obstacles, so leaving that up to robot (or fleet manager, depending on the design I suppose) would make for really great synergy between the fleet adapter and the fleet manager. We've had great success with this approach on the MiR.

If, for some reason, the fleet manager **must** have complete control over the entire path from start to end, with no ability to specify intermediary waypoints or to feed in waypoints during task execution, then it would not qualify for the top-tier of [RMF fleet compliance](https://osrf.github.io/ros2multirobotbook/rmf-core.html#fleet-adapters): the **Full Control** category. Instead it would fall under the **Read Only** category (or **Traffic Light** if it can at least start or stop along the way, but we don't have a user-friendly API for that yet, so you'd have to implement the logic for that if you need it soon).

> Will such discarding of path information affect the internal operation of rmf_core?

Discarding the path information and trying to go directly to the destination is not an acceptable way to use the `rmf_fleet_adapter::agv` API. The fleet adapter reports the path that it planned to the schedule, so if your robot follows a different path, then it's not complying with the schedule at all. You would be better off writing your own fleet adapter from scratch using `rmf_traffic_ros2` and `rmf_traffic_msgs`. I don't recommend doing this because, frankly, it's very difficult.

I should also note, when interfacing with the `rmf_fleet_adapter::agv` API, you should not just follow the waypoints that are given, but you should also be conscientious of the timing on those waypoints. Specifically, you should not pass a given waypoint until the clock has passed that waypoint's time value. This timing information is a crucial aspect of the fleet adapter's strategy for avoiding conflicts, because it may have your vehicle wait at a spot for some period of time in order to allow another robot to pass. So not only would it be wrong to discard the path, but it would also be wrong to discard the timing information. Note that it's perfectly okay for your robot to be late in arriving at a waypoint, but it must not leave a waypoint until the designated time, or else it is likely to create conflicts that should have been avoided.

### In summary

If your fleet manager's API allows it, I would strongly advise feeding the path waypoints to it one at a time as a series of `DestinationRequest`s. Do not send the next `DestinationRequest` until **both** conditions are met:

1. The robot has arrived at the current waypoint
2. The clock has reached or gone past the time specified for the current waypoint

**Comment by rcywongaa** [comment]
I think my situation falls under the **Read Only** category (technically **Traffic Light** but since it's not available at the moment, I'll treat it as **Read Only** for now).

In that case, what is the procedure for implementing a read-only fleet adapter?

The [`rmf_fleet_adapter/src/read_only/FleetAdapterNode.hpp`](https://github.com/osrf/rmf_core/blob/4099921bea576c2fde87c3dc852cc4d8f86a5d1a/rmf_fleet_adapter/src/read_only/FleetAdapterNode.hpp) doesn't seem like it supports inheriting or extending.

**Comment by mxgrey** [comment]
Before proceeding down that path, are you certain it is impossible to incrementally issue a sequence of `DestinationRequest`s to your fleet manager? If you're working with Sesto, I believe that was successfully done with the old fleet adapter+driver API, so it should be entirely possible with the new API.

This basically means you issue a `DestinationRequest` that targets the first waypoint given by `follow_new_path` and then wait until the robot arrives before issuing the next `DestinationRequest` to the next waypoint... etc until the robot reaches the final waypoint.

I want to make sure that is truly not an option before proceeding, because it would be a shame to needlessly demote a robot.

**Comment by mxgrey** [comment]
The traffic light API has now been merged as of #147 .

Since the initial issue here was the absence of the traffic light API, I'll close this ticket.
